File: lib/model/global_app_state.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/cupertino.dart';
import 'package:flutter_gen/gen_l10n/app_localizations.dart';
import 'package:go_list/model/collections/shopping_list_collection.dart';
import 'package:go_list/model/item.dart';
import 'package:go_list/model/settings.dart';
import 'package:go_list/model/shopping_list.dart';
import 'package:go_list/service/golist_client.dart';
import 'package:go_list/service/golist_languages.dart';
import 'package:go_list/service/items/input_to_item_parser.dart';
import 'package:go_list/service/storage/provider/local_storage_provider.dart';
import 'package:go_list/service/storage/provider/remote_storage_provider.dart';
import 'package:go_list/service/storage/shopping_list_storage.dart';
import 'package:go_list/view/shopping_list/shopping_list_item/shopping_list_item.dart';

class GlobalAppState extends ChangeNotifier {
  final GoListClient goListClient;
  late final ShoppingListStorage shoppingListStorage;
  late ShoppingListCollection shoppingLists;
  late Settings settings;
  StreamSubscription<ShoppingList>? _streamSubscription;
  bool shouldShowConnectionFailure = false;
  Timer? _showConnectionFailureTimer;

  // for undo feature
  List<String> recentlyDeletedItems = [];
  Map<String, Timer> removeRecentlyDeletedItemTimers = {};

  GlobalAppState(
      {required this.goListClient,
      required LocalStorageProvider localStorageProvider,
      required RemoteStorageProvider remoteStorageProvider}) {
    shoppingListStorage =
        ShoppingListStorage(localStorageProvider, remoteStorageProvider);
    shoppingLists = shoppingListStorage.loadShoppingListsFromLocalStorage();

    // load settings or create default
    Settings? settingsFromStorage = shoppingListStorage.loadSettings();
    if (settingsFromStorage == null ||
        !shoppingLists
            .containsEntryWithId(settingsFromStorage.selectedShoppingListId)) {
      settings = Settings(
          selectedShoppingListId: shoppingLists.first()?.id ?? "",
          shoppingListOrder: shoppingLists.order);
      shoppingListStorage.saveSettings(settings);
    } else {
      settings = settingsFromStorage;
    }

    shoppingLists.setOrder(settings.shoppingListOrder);
    goListClient.deviceId = settings.deviceId;
    InputToItemParser().init(languageCode);
    _initWithDefaultListIfEmpty();
    _listenForChangesInSelectedShoppingList();
    notifyListeners();
  }

  ShoppingList? _initWithDefaultListIfEmpty() {
    if (shoppingLists.length == 0) {
      ShoppingList defaultList = ShoppingList(name: "");
      shoppingLists.upsert(defaultList);
      shoppingListStorage.upsertShoppingList(defaultList);
      settings.selectedShoppingListId = defaultList.id;
      shoppingListStorage.saveSettings(settings);
      AppLocalizations.delegate.load(settings.locale).then((appLocalizations) {
        defaultList.name = appLocalizations.default_name;
        shoppingListStorage.upsertShoppingList(defaultList);
      });
      return defaultList;
    }
    return null;
  }

  Future<void> loadListsFromStorage() async {
    await shoppingListStorage
        .loadShoppingLists()
        .listen(setShoppingLists)
        .asFuture()
        .onError((e, s) {
      showConnectionFailure();
    });
    _listenForChangesInSelectedShoppingList();
  }

  void showConnectionFailure() {
    shouldShowConnectionFailure = true;
    print("shouldShowConnectionFailure $shouldShowConnectionFailure");
    notifyListeners();
    _showConnectionFailureTimer?.cancel();
    _showConnectionFailureTimer = Timer(const Duration(seconds: 5), () {
      shouldShowConnectionFailure = false;
      print("shouldShowConnectionFailure $shouldShowConnectionFailure");
      notifyListeners();
    });
  }

  setShoppingLists(ShoppingListCollection shoppingLists) {
    if (!this.shoppingLists.equals(shoppingLists)) {
      this.shoppingLists = shoppingLists;
      this.shoppingLists.setOrder(settings.shoppingListOrder);
      notifyListeners();
    }
  }

  setShoppingListOrder(List<String> shoppingListOrder) {
    shoppingLists.setOrder(shoppingListOrder);
    settings.shoppingListOrder = shoppingListOrder;
    shoppingListStorage.saveSettings(settings);
    notifyListeners();
  }

  void setSelectedShoppingListId(String selectedShoppingListId) {
    settings.selectedShoppingListId = selectedShoppingListId;
    shoppingListStorage.saveSettings(settings);
    notifyListeners();
    _listenForChangesInSelectedShoppingList(forceReconnect: true);
  }

  void _listenForChangesInSelectedShoppingList(
      {forceReconnect = false, retries = 0}) async {
    onError(e) {
      print("failed to listen for changes: $e");
      _streamSubscription = null;
      if (retries == 0) {
        showConnectionFailure();
      }
    }

    try {
      if (forceReconnect) {
        await _streamSubscription?.cancel();
        _streamSubscription = null;
      }
      if (_streamSubscription != null || retries > 5) return;
      _streamSubscription = (await shoppingListStorage
              .listenForChanges(settings.selectedShoppingListId))
          .listen((shoppingList) {
        shoppingLists.upsert(shoppingList);
        notifyListeners();
      });
      _streamSubscription!.onDone(() {
        _streamSubscription = null;
        _listenForChangesInSelectedShoppingList(retries: retries + 1);
      });
      _streamSubscription!.onError(onError);
    } on SocketException catch (e) {
      onError(e);
    }
  }

  void deleteShoppingList(String shoppingListId) {
    shoppingLists.removeEntryWithId(shoppingListId);
    _initWithDefaultListIfEmpty();
    // if selected list was deleted select the first list
    if (settings.selectedShoppingListId == shoppingListId) {
      settings.selectedShoppingListId = shoppingLists.first()!.id;
      shoppingListStorage.saveSettings(settings);
    }
    shoppingListStorage.deleteShoppingList(shoppingListId);
    notifyListeners();
  }

  void upsertShoppingList(ShoppingList shoppingList) {
    shoppingLists.upsert(shoppingList);
    shoppingListStorage.upsertShoppingList(shoppingList);
    setSelectedShoppingListId(shoppingList.id);
    notifyListeners();
  }

  void updateListOrder(int oldIndex, int newIndex) {
    shoppingLists.moveEntryInOrder(oldIndex, newIndex);
    settings.shoppingListOrder = shoppingLists.order;
    shoppingListStorage.saveSettings(settings);
    notifyListeners();
  }

  ShoppingList get selectedShoppingList =>
      shoppingLists.entryWithId(settings.selectedShoppingListId)!;

  void deleteItem(Item item) {
    selectedShoppingList.deleteItem(item);
    shoppingListStorage.upsertShoppingList(selectedShoppingList);
    recentlyDeletedItems.add(item.id);
    removeRecentlyDeletedItemTimers[item.id]?.cancel();
    removeRecentlyDeletedItemTimers[item.id] = Timer(
        const Duration(milliseconds: ShoppingListItem.allowUndoForMs), () {
      recentlyDeletedItems.remove(item.id);
      notifyListeners();
    });
    notifyListeners();
  }

  void unDeleteItem(String itemId) {
    selectedShoppingList.unDeleteItem(itemId);
    shoppingListStorage.upsertShoppingList(selectedShoppingList);
    recentlyDeletedItems.remove(itemId);
    removeRecentlyDeletedItemTimers[itemId]?.cancel();
    removeRecentlyDeletedItemTimers.remove(itemId);
    print("undelete");
    notifyListeners();
  }

  void upsertItem(Item item) {
    selectedShoppingList.upsertItem(item);
    shoppingListStorage.upsertShoppingList(selectedShoppingList);
    notifyListeners();
  }

  setLocale(Locale locale) {
    settings.language = locale.languageCode;
    shoppingListStorage.saveSettings(settings);
    InputToItemParser().init(languageCode);
    notifyListeners();
  }

  Locale get locale => Locale(languageCode);

  String get languageName => GoListLanguages.getLanguageName(settings.language);

  String get languageCode => settings.language;
}

File: svelte-golist/src/lib/stores/globalAppState.ts
import { writable, derived, get } from 'svelte/store';
import { browser } from '$app/environment';
import { ShoppingList } from '../models/ShoppingList.js';
import { ShoppingListCollection } from '../models/collections/ShoppingListCollection.js';
import { Settings } from '../models/Settings.js';
import { Item } from '../models/Item.js';
import { GoListClient } from '../services/GoListClient.js';
import { LocalStorageProvider } from '../services/storage/LocalStorageProvider.js';
import { RemoteStorageProvider } from '../services/storage/RemoteStorageProvider.js';
import { ShoppingListStorage } from '../services/storage/ShoppingListStorage.js';
import { InputToItemParser } from '../services/InputToItemParser.js';

// Core stores
export const shoppingLists = writable<ShoppingListCollection>(new ShoppingListCollection([]));
export const settings = writable<Settings | null>(null);
export const shouldShowConnectionFailure = writable<boolean>(false);
export const recentlyDeletedItems = writable<string[]>([]);

// Services
let goListClient: GoListClient;
let shoppingListStorage: ShoppingListStorage;
let removeRecentlyDeletedItemTimers: Map<string, number> = new Map();
let showConnectionFailureTimer: number | null = null;
let streamReader: ReadableStreamDefaultReader<ShoppingList> | null = null;

// Derived stores
export const selectedShoppingList = derived(
  [shoppingLists, settings],
  ([$shoppingLists, $settings]) => {
    if (!$settings) return null;
    return $shoppingLists.entryWithId($settings.selectedShoppingListId);
  }
);

export const languageCode = derived(settings, ($settings) => $settings?.language || 'en');

export const locale = derived(languageCode, ($languageCode) => $languageCode);

// Initialize the app state
export async function initializeApp(): Promise<void> {
  if (!browser) return;

  await LocalStorageProvider.init();
  const localStorageProvider = new LocalStorageProvider();
  await localStorageProvider.migrateFromPreviousVersion();

  goListClient = new GoListClient();
  const remoteStorageProvider = new RemoteStorageProvider(goListClient);
  shoppingListStorage = new ShoppingListStorage(localStorageProvider, remoteStorageProvider);

  const shoppingListsFromStorage = shoppingListStorage.loadShoppingListsFromLocalStorage();
  shoppingLists.set(shoppingListsFromStorage);

  // Load or create settings
  let settingsFromStorage = shoppingListStorage.loadSettings();
  if (!settingsFromStorage || !shoppingListsFromStorage.containsEntryWithId(settingsFromStorage.selectedShoppingListId)) {
    settingsFromStorage = new Settings({
      selectedShoppingListId: shoppingListsFromStorage.first()?.id || '',
      shoppingListOrder: shoppingListsFromStorage.order
    });
    shoppingListStorage.saveSettings(settingsFromStorage);
  }

  settings.set(settingsFromStorage);
  shoppingListsFromStorage.setOrder(settingsFromStorage.shoppingListOrder);
  goListClient.deviceId = settingsFromStorage.deviceId;

  // Initialize input parser
  await InputToItemParser.getInstance().init(settingsFromStorage.language);

  // Create default list if empty
  initWithDefaultListIfEmpty();
  
  // Start listening for changes
  listenForChangesInSelectedShoppingList();
}

function initWithDefaultListIfEmpty(): void {
  const currentShoppingLists = get(shoppingLists);
  const currentSettings = get(settings);
  
  if (currentShoppingLists.length === 0) {
    const defaultList = new ShoppingList({ name: 'Shopping List' });
    currentShoppingLists.upsert(defaultList);
    shoppingListStorage.upsertShoppingList(defaultList);
    
    if (currentSettings) {
      currentSettings.selectedShoppingListId = defaultList.id;
      shoppingListStorage.saveSettings(currentSettings);
      settings.set(currentSettings);
    }
    
    shoppingLists.set(currentShoppingLists);
  }
}

export async function loadListsFromStorage(): Promise<void> {
  try {
    for await (const updatedShoppingLists of shoppingListStorage.loadShoppingLists()) {
      shoppingLists.set(updatedShoppingLists);
    }
    listenForChangesInSelectedShoppingList();
  } catch (error) {
    console.error('Failed to load lists from storage:', error);
    showConnectionFailure();
  }
}

function showConnectionFailure(): void {
  shouldShowConnectionFailure.set(true);
  
  if (showConnectionFailureTimer) {
    clearTimeout(showConnectionFailureTimer);
  }
  
  showConnectionFailureTimer = setTimeout(() => {
    shouldShowConnectionFailure.set(false);
  }, 5000);
}

export function setShoppingListOrder(order: string[]): void {
  const currentShoppingLists = get(shoppingLists);
  const currentSettings = get(settings);
  
  if (currentSettings) {
    currentShoppingLists.setOrder(order);
    currentSettings.shoppingListOrder = order;
    shoppingListStorage.saveSettings(currentSettings);
    shoppingLists.set(currentShoppingLists);
    settings.set(currentSettings);
  }
}

export function setSelectedShoppingListId(selectedShoppingListId: string): void {
  const currentSettings = get(settings);
  
  if (currentSettings && selectedShoppingListId !== currentSettings.selectedShoppingListId) {
    recentlyDeletedItems.set([]);
    removeRecentlyDeletedItemTimers.clear();
    
    currentSettings.selectedShoppingListId = selectedShoppingListId;
    shoppingListStorage.saveSettings(currentSettings);
    settings.set(currentSettings);
    
    listenForChangesInSelectedShoppingList(true);
  }
}

async function listenForChangesInSelectedShoppingList(forceReconnect = false): Promise<void> {
  const currentSettings = get(settings);
  if (!currentSettings) return;

  try {
    if (forceReconnect && streamReader) {
      await streamReader.cancel();
      streamReader = null;
    }

    if (streamReader) return;

    const stream = await shoppingListStorage.listenForChanges(currentSettings.selectedShoppingListId);
    streamReader = stream.getReader();

    const pump = async () => {
      try {
        while (true) {
          const { done, value } = await streamReader!.read();
          if (done) break;
          
          const currentShoppingLists = get(shoppingLists);
          currentShoppingLists.upsert(value);
          shoppingLists.set(currentShoppingLists);
        }
      } catch (error) {
        console.error('Stream error:', error);
        showConnectionFailure();
      } finally {
        streamReader = null;
        // Attempt to reconnect after a delay
        setTimeout(() => listenForChangesInSelectedShoppingList(), 5000);
      }
    };

    pump();
  } catch (error) {
    console.error('Failed to listen for changes:', error);
    showConnectionFailure();
  }
}

export function deleteShoppingList(shoppingListId: string): void {
  const currentShoppingLists = get(shoppingLists);
  const currentSettings = get(settings);
  
  currentShoppingLists.removeEntryWithId(shoppingListId);
  initWithDefaultListIfEmpty();
  
  if (currentSettings && currentSettings.selectedShoppingListId === shoppingListId) {
    const firstList = currentShoppingLists.first();
    if (firstList) {
      currentSettings.selectedShoppingListId = firstList.id;
      shoppingListStorage.saveSettings(currentSettings);
      settings.set(currentSettings);
    }
  }
  
  shoppingListStorage.deleteShoppingList(shoppingListId);
  shoppingLists.set(currentShoppingLists);
}

export function upsertAndSelectShoppingList(shoppingList: ShoppingList): void {
  const currentShoppingLists = get(shoppingLists);
  currentShoppingLists.upsert(shoppingList);
  shoppingListStorage.upsertShoppingList(shoppingList);
  setSelectedShoppingListId(shoppingList.id);
  shoppingLists.set(currentShoppingLists);
}

export function deleteItem(itemId: string): void {
  const currentSelectedList = get(selectedShoppingList);
  if (!currentSelectedList) return;

  currentSelectedList.deleteItem(itemId);
  shoppingListStorage.upsertShoppingList(currentSelectedList);
  
  const currentRecentlyDeleted = get(recentlyDeletedItems);
  recentlyDeletedItems.set([...currentRecentlyDeleted, itemId]);
  
  // Clear existing timer for this item
  const existingTimer = removeRecentlyDeletedItemTimers.get(itemId);
  if (existingTimer) {
    clearTimeout(existingTimer);
  }
  
  // Set new timer to remove from recently deleted
  const timer = setTimeout(() => {
    const currentRecentlyDeleted = get(recentlyDeletedItems);
    recentlyDeletedItems.set(currentRecentlyDeleted.filter(id => id !== itemId));
    removeRecentlyDeletedItemTimers.delete(itemId);
  }, 5000); // 5 seconds to undo
  
  removeRecentlyDeletedItemTimers.set(itemId, timer);
  
  // Trigger reactivity
  const currentShoppingLists = get(shoppingLists);
  shoppingLists.set(currentShoppingLists);
}

export function unDeleteItem(): void {
  const currentRecentlyDeleted = get(recentlyDeletedItems);
  const currentSelectedList = get(selectedShoppingList);
  
  if (currentRecentlyDeleted.length === 0 || !currentSelectedList) return;

  const lastDeletedItemId = currentRecentlyDeleted[0];
  currentSelectedList.unDeleteItem(lastDeletedItemId);
  shoppingListStorage.upsertShoppingList(currentSelectedList);
  
  recentlyDeletedItems.set(currentRecentlyDeleted.filter(id => id !== lastDeletedItemId));
  
  const timer = removeRecentlyDeletedItemTimers.get(lastDeletedItemId);
  if (timer) {
    clearTimeout(timer);
    removeRecentlyDeletedItemTimers.delete(lastDeletedItemId);
  }
  
  // Trigger reactivity
  const currentShoppingLists = get(shoppingLists);
  shoppingLists.set(currentShoppingLists);
}

export function upsertItem(item: Item): void {
  const currentSelectedList = get(selectedShoppingList);
  if (!currentSelectedList) return;

  currentSelectedList.upsertItem(item);
  shoppingListStorage.upsertShoppingList(currentSelectedList);
  
  // Trigger reactivity
  const currentShoppingLists = get(shoppingLists);
  shoppingLists.set(currentShoppingLists);
}

export function setLocale(newLocale: string): void {
  const currentSettings = get(settings);
  if (!currentSettings) return;

  currentSettings.language = newLocale;
  shoppingListStorage.saveSettings(currentSettings);
  InputToItemParser.getInstance().init(newLocale);
  settings.set(currentSettings);
}